tpl: Add partialCached template function

moorereason opened the pull request:
Fixes #1368

moorereason commented:
@bep, can  you double-check my test code?  I used `InitializeT()`, but I'm not sure if that's the proper way to setup the template system.  Without it, I was getting a panic in `Lookup()` since `tmpl` is nil.

I would have copied the partial tests, but we don't have any!  :)

moorereason commented:
@bep, just noticed your proposed API.  I'll take a look at updating my PR.

bep commented:
Yes, my thought was to make it possible to use the same partial with different contexts without making a copy of the partial template.

bep commented on tpl/template_funcs.go:
```diff
@@ -1392,6 +1392,41 @@ func replace(a, b, c interface{}) (string, error) {
+
+	if p, ok = c.get(name); !ok {
+		p = partial(name, contextList...)
+		c.set(name, p)
```
It may not have much practical difference in this case, but this lock construction with a `set` and `get` method makes it impossible to get the locking right -- and since people are going to copy this code into their university coding assignments, we might as well do it right.

* The creation of the partial isn't protected by a lock, so there will be multiple invocations of the same `partial`
* Also it would be good to double-check the map once you get the write lock; the partial may have been added in the open window between unlock and lock.

moorereason commented:
@bep, I updated the locking code.  I agree with your original comment.  I've moved the partial rendering into the `set` func.  Also reworked to support an optional variant parameter.

If you like what you see, I'll add some docs to the PR prior to merging.

bep commented on tpl/template_funcs.go:
```diff
@@ -1392,6 +1392,47 @@ func replace(a, b, c interface{}) (string, error) {
+var cachedPartials = partialCache{p: make(map[string]template.HTML)}
+
+func partialCached(name string, context interface{}, variant ...string) template.HTML {
+	key := name + strings.Join(variant, "")
```
This is nitpicky, but I have seen some fairly heavy use of partials lately, so this would be better:

```go
key := name
if len(variant) > 0 {
   key += ...
}

bep commented on tpl/template_funcs.go:
```diff
@@ -1392,6 +1392,47 @@ func replace(a, b, c interface{}) (string, error) {
+	return
+}
+
+func (c *partialCache) set(key, name string, context interface{}) template.HTML {
```
There is still a window between get/set where the partial could have been created by others. That is why I commented on the get/set construction. You can fix it by adding an additional map lookup inside the write lock, but then the set/get method names will be slightly misleading. As such I prefer reading the whole thing as one big code block. You would still need a double-check inside the write lock, but mistakes will be easier to spot.

Note that in normal cache situations (like a regexp cache), I may leave out the double-check because it doesn't matter if it gets invoked 1 or 2 times -- but in this case it has potential to get unexpected results (different partials).

bep commented on tpl/template_funcs.go:
```diff
@@ -1392,6 +1392,44 @@ func replace(a, b, c interface{}) (string, error) {
+func (c *partialCache) Get(key, name string, context interface{}) (p template.HTML) {
+	var ok bool
+
+	c.mu.Lock()
```
By double-checking i meant ... double. In this case you could use a regular mutex -- but I'm pretty sure proper use of a rwmutex will get much faster in this case.

So:

```
RLock
p, ok = m[key]
RUnlock

if ok return p

Lock

// double check
if p, ok = m[key]

else p = partial ...
Unlock

```

bep commented on tpl/template_funcs.go:
```diff
@@ -1392,6 +1392,44 @@ func replace(a, b, c interface{}) (string, error) {
 
+// partialCache represents a cache of partials protected by a mutex.
+type partialCache struct {
+	mu sync.RWMutex
```
I prefer to embed the mutex when the type is unexported and the mutex is so connected to the type, then you can do `c.Lock()` etc.

moorereason commented:
Travis was hung, so I squashed and rebased.
